tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+}
+
+static pg_attribute_always_inline void
+get_key_arrow_fixed_8(CompressedColumnValues column, int row, Datum *restrict key,
+					  bool *restrict valid)
+{
+#ifndef USE_FLOAT8_BYVAL
+	/*
+	 * Shouldn't be called for this configuration, because we only use this
+	 * grouping strategy for by-value types.
+	 */
+	Assert(false);
+#endif
+
+	get_key_arrow_fixed(column, row, 8, key, valid);
+}
+
+/*
+ * Implementation of bulk hashing specialized for a given key width.
+ */
+static pg_noinline uint32
+fill_offsets_arrow_fixed_8(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_8);
+}
+
+static pg_noinline uint32
+fill_offsets_arrow_fixed_4(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_4);
+}
+
+static pg_noinline uint32
+fill_offsets_arrow_fixed_2(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+						   int key_column_index, uint32 next_unused_state_index, int start_row,
+						   int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_arrow_fixed_2);
+}
+
+static pg_noinline uint32
+fill_offsets_scalar(GroupingPolicyHash *policy, DecompressBatchState *batch_state,
+					int key_column_index, uint32 next_unused_state_index, int start_row,
+					int end_row, uint32 *restrict offsets)
+{
+	return fill_offsets_dispatch(policy,
+								 batch_state,
+								 key_column_index,
+								 next_unused_state_index,
+								 start_row,
+								 end_row,
+								 offsets,
+								 get_key_scalar);
+}
+
+static void
+gp_hash_add_batch(GroupingPolicy *gp, DecompressBatchState *batch_state)
+{
+	GroupingPolicyHash *policy = (GroupingPolicyHash *) gp;
+
+	Assert(!policy->returning_results);
+
+	const uint64_t *restrict filter = batch_state->vector_qual_result;
+	const int n = batch_state->total_batch_rows;
+
+	/*
+	 * Initialize the array for storing the aggregate state offsets corresponding
+	 * to a given batch row.
+	 */
+	if ((size_t) n > policy->num_allocated_offsets)
+	{
+		policy->num_allocated_offsets = n;
+		policy->offsets = palloc(sizeof(policy->offsets[0]) * policy->num_allocated_offsets);
```

Comment:
We don't use repalloc because the data is per-batch and we don't care about the old values. I also added a pfree. The context is the ExecutorState where the vector_agg_exec is performed, as with most other allocations here.